### PR TITLE
Throughput test late unlock [10237]

### PR DESCRIPTION
**Pull request by jparisu**
Change unlock order in code to try to fix unknown failure in ThroughputTest.

Log added to keep tracking in next failure.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3516/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/706/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3126/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3548/)

File: test/performance/throughput/ThroughputPublisher.cpp [hunk 1]
```diff
@@ -70,11 +70,12 @@ void ThroughputPublisher::DataPubListener::onPublicationMatched(
         --throughput_publisher_.data_discovery_count_;
     }
 
+    lock.unlock();
+
     if (throughput_publisher_.data_discovery_count_ == static_cast<int>(throughput_publisher_.subscribers_))
     {
         throughput_publisher_.data_discovery_cv_.notify_one();
     }
```
Comment (EduPonz) [inline test/performance/throughput/ThroughputPublisher.cpp:78]: I'm not sure whether unlocking and the checking the count is a good idea. We could do something like:

```suggestion
    if (throughput_publisher_.data_discovery_count_ == static_cast<int>(throughput_publisher_.subscribers_))
    {
        lock.unlock();
        throughput_publisher_.data_discovery_cv_.notify_one();
        return;
    }
    lock.unlock();
```

File: test/performance/throughput/ThroughputPublisher.cpp [hunk 2]
```diff
@@ -108,12 +109,12 @@ void ThroughputPublisher::CommandSubListener::onSubscriptionMatched(
         --throughput_publisher_.command_discovery_count_;
     }
 
+    lock.unlock();
+
```
Comment (EduPonz) [inline test/performance/throughput/ThroughputPublisher.cpp:113]: Same applies here

File: test/performance/throughput/ThroughputPublisher.cpp [hunk 3]
```diff
@@ -148,12 +149,12 @@ void ThroughputPublisher::CommandPubListener::onPublicationMatched(
         --throughput_publisher_.command_discovery_count_;
     }
 
-    if (throughput_publisher_.command_discovery_count_ == static_cast<int>(throughput_publisher_.subscribers_ * 2) )
+    lock.unlock();
+
```
Comment (EduPonz) [inline test/performance/throughput/ThroughputPublisher.cpp:153]: Same here

**Review by EduPonz — CHANGES_REQUESTED** [review]

File: test/performance/throughput/ThroughputPublisher.cpp (hunk 1 above)
Comment (jparisu) [reply test/performance/throughput/ThroughputPublisher.cpp]: :+1:

File: test/performance/throughput/ThroughputPublisher.cpp (hunk 2 above)
Comment (jparisu) [reply test/performance/throughput/ThroughputPublisher.cpp]: :+1:

File: test/performance/throughput/ThroughputPublisher.cpp (hunk 3 above)
Comment (jparisu) [reply test/performance/throughput/ThroughputPublisher.cpp]: :+1:

**Review by EduPonz — APPROVED** [review]
LGTM, waiting on CI

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3520/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/710/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3130/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3552/)

**Pull request merged**